### PR TITLE
Add support for general non-necessarily-recursive binders in notations

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

We provide a new class of (non-recursive) binders for notations supporting all atomic forms of Coq binders: `a`, `{a}`, `[a]`, `a:T`, `{a:T}`, `[a:T]`, `'pat` (formerly, we were accepting either names or patterns). In particular, the implicit arguments are taken into account.

We accept also the closed form of binders which differs in `a:T` being replaced by `(a:T)`.

Example:
```
Notation "## x P" := (forall x, x = x -> P) (x binder, at level 0).
Check ## '(x,y) (x+y=0).
Check fun (f : ## {a} (a=0)) => f (a:=1) eq_refl.
```

In the process, we slightly extend the support for casts in pattern. For instance, the following is now accepted:
```
Notation "## x P" := (forall x, P) (x pattern, at level 0).
Check ## ((x,y):bool*bool) (x=y).
```

This was initially suggested on Iris mattermost by @robbertkrebbers.

- [X] Added / updated test-suite
- [ ] Extra stress tests
- [X] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog

### CODE REVIEW

**Comment by robbertkrebbers** [comment]
Many thanks!

> `Check ## ((x,y):bool*bool) (x=y).`

I am a bit confused here, normally when you have a pattern matching lambda, you need a `'`, but it seems like that in combination with a type cast that's not needed? Why is that so?

**Comment by herbelin** [comment]
> > `Check ## ((x,y):bool*bool) (x=y).`
> 
> I am a bit confused here, normally when you have a pattern matching lambda, you need a `'`, but it seems like that in combination with a type cast that's not needed? Why is that so?

The `'` is not part of the grammar of `pattern` itself, as the grammar of `pattern` is per se independent of the grammar of binders. For instance, it is used in `match ... with ... end` and it does not take a `'` there.

This is not related to the question of type casts. Pattern type casts are part of the syntax of patterns, so you have them as soon as you parse a `pattern`. The `'` is only part of the syntax embedding `pattern` into `binder` (i.e., `binder ::= _ | ident | ' pattern | ident : type | ...`). Does it answer?

**Comment by robbertkrebbers** [comment]
> The `'` is not part of the grammar of `pattern` itself, as the grammar of `pattern` is per se independent of the grammar of binders. For instance, it is used in `match ... with ... end` and it does not take a `'` there.
> 
> This is not related to the question of type casts. Pattern type casts are part of the syntax of patterns, so you have them as soon as you parse a `pattern`. The `'` is only part of the syntax embedding `pattern` into `binder` (i.e., `binder ::= _ | ident | ' pattern | ident : type | ...`). Does it answer?

That helps, but how does `((x,y):bool*bool)` fit into the grammar of `binder` you just gave?

**Comment by herbelin** [comment]
> That helps, but how does `((x,y):bool*bool)` fit into the grammar of binder you just gave?

Sorry, I missed this question. The expression `((x,y):bool*bool)` (without quote) is part of the syntax of patterns (case of a typed pattern).

**Comment by herbelin** [comment]
I believe I can now turn this PR to ready-for-review.

File: parsing/g_constr.mlg [hunk 1]
```diff
@@ -438,13 +443,20 @@ GRAMMAR EXTEND Gram
...
-      | "'"; p = pattern LEVEL "0" ->
-        { let (p, ty) =
-            match p.CAst.v with
-            | CPatCast (p, ty) -> (p, Some ty)
-            | _ -> (p, None)
-          in
-          [CLocalPattern (CAst.make ~loc (p, ty))] } ] ]
+      | "'"; p = pattern LEVEL "0" -> { [CLocalPattern p] } ] ]
+  ;
+  one_open_binder:
```
Comment (jfehrle) [inline parsing/g_constr.mlg:448]: I didn't see where one_open_binder or one_closed_binder are used.  Would you run `make doc_gram_rsts` and include those productions in an appropriate place in the manual?

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 2]
```diff
@@ -689,6 +689,31 @@ Note also that in the absence of a ``as ident``, ``as strict pattern`` or
 ``as pattern`` :n:`@syntax_modifier`\s, the default is to consider sub-expressions occurring
 in binding position and parsed as terms to be ``as ident``.
 
+Binders bound in the notation and parsed as general binders
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:692]: Even after 10 minutes I still don't understand this--what's been added, when it is useful.  The existing description of Notation and syntax_modifier need considerable work (separate from this PR).  Hope you can make this better.

**Review by ejgallego — APPROVED** [review]
Looks good to me in principle, but I agree that the documentation could be tweaked a bit more.

We should really take some time to document the substitution rules of notations a bit.

For example, what is this supposed to do [writing in pseudo-code]:
```coq
Notation "## x P" := (forall {x}, P x) (binder x).
Check (## [x] (fun x => True)).
```

File: parsing/g_constr.mlg (hunk 1 above)
Comment (herbelin) [reply parsing/g_constr.mlg:448]: They are called in `vernac/egramcoq.ml`. They are supposed to be visible only with `Print Grammar constr` (or `Print Custom Grammar foo`) for the notations having the `binder` modifier. They are of somehow internal use and it is not clear whether it is important to document them.

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 2 above)
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst:692]: There is two levels of binding. The bindings in terms, like in `fun`, `forall`, `let` and the bindings of subterm expressions by variables of a notation. The second are at a meta-level since they operate on terms which themselves have binders at the (lower) term-level (also called object-level in logic).

In particular, one can bind at the meta-level variables which play a role of binder at the term-level. Otherwise said, "Binders bound" should be read as "Term-binders bound at the notation level".

**Comment by herbelin** [comment]
> We should really take some time to document the substitution rules of notations a bit.

Let me consider a slightly working and more general example:
```
Notation "## x P" := (forall {x} y, P x x y) (x binder, at level 0).
Check fun f : ## [x] (fun x y t => x = y + t) => f 0.
```

There is a bit of documentation in the [Notation and binders](https://coq.inria.fr/refman/user-extensions/syntax-extensions.html#notations-and-binders) section but not very formal. We could probably give more examples.

In the above example, the actual `x` is substituted as a pattern to the three occurrences of `x` in the pattern (i.e. once as a binding variable in `{x}` and twice translated as a term in the `x`s of `P x x y`). Also, `P` being in the scope of `{x}`, the variables of the pattern substituted to `{x}` are binding in `P`.

The `x` of the `fun x` is unrelated (i.e. binding variables of the actual instance are not meta-binding-variables), so it binds the `x` in `x = y + t` but shadows the effect of substituting the actual `x` to `{x}` in the pattern.

The `y` in the pattern is not bound on the left, so it is not a meta-binding-variable. It just behaves like an ordinary binder (up to alpha-conversion).

The `y` in the actual term also behaves like an ordinary binder (up to alpha-conversion).

The effect of brackets in `{x}` is the pattern lost (there is an issue somewhere about that). But the`[x]` has an effect on the implicit arguments of `f`.

In short, this is elaborated into `fun f : forall x y, (fun x0 y0 t => x0 = y0 + t) x x y => f _ 0`.

**Comment by jfehrle** [comment]
How about adding something like this as an example in the doc?  I'd suggest avoiding reusing identifiers to make it easier to follow (which x is which?), and possibly splitting it into 2 separate examples.  I think it's best to make the examples small enough that the points are unmistakable.

**Comment by robbertkrebbers** [comment]
> > That helps, but how does `((x,y):bool*bool)` fit into the grammar of binder you just gave?
> 
> Sorry, I missed this question. The expression `((x,y):bool*bool)` (without quote) is part of the syntax of patterns (case of a typed pattern).

I see. Is this grammar, in one way or another, included in the documentation somewhere? If so, it would also be good to have some examples, like:

```coq
Definition foo1 := fun xy : nat*bool => xy.      (* Works *)
Definition foo2 := fun '((x,y) : nat*bool) => x. (* Works *)
Definition foo3 := fun '(x,y) : nat*bool => x.   (* Doesn't work *)
```

**Comment by herbelin** [comment]
>  Is this grammar, in one way or another, included in the documentation somewhere?

The syntax is described [there](https://coq.inria.fr/refman/language/core/assumptions.html#functions-and-assumptions), which indicates, following the `pattern0` link, that parentheses are needed and that `(pat : typ)` is not unavailable for `fun 'pat`.

I agree that some examples would be good, which could typically go in the section I mentioned.

**Comment by ejgallego** [comment]
> In short, this is elaborated into `fun f : forall x y, (fun x0 y0 t => x0 = y0 + t) x x y => f _ 0`.

Thanks for the explanation @herbelin , is there an intermediate term contructed in intern , or it just jumps to this form directly.

Unfortunately this needs a rebase, will merge afterwards, unless you want to tweak docs.

**Comment by herbelin** [comment]
> Unfortunately this needs a rebase, will merge afterwards, unless you want to tweak docs.

If you give me a few days, I'll add an example in the doc.

**Comment by ejgallego** [comment]
> If you give me a few days, I'll add an example in the doc.

Sounds good @herbelin , but I think I'll merge this sooner given the target milestone, doing the documentation PR should be simpler for @gares to backport, also we want testing sooner than later.

**Comment by Zimmi48** [comment]
For now the example in the doc doesn't even build though.

**Comment by herbelin** [comment]
> >  If you give me a few days, I'll add an example in the doc.

> Sounds good @herbelin , but I think I'll merge this sooner given the target milestone, doing the documentation PR should be simpler for @gares to backport, also we want testing sooner than later.

OK, then I'll refine then doc on binders later, probably at the time of addressing the generalization mentioned in #12467.

I rebased and fixed the refman example (hopefully...).

**Comment by Zimmi48** [comment]
If it's about adding examples of already existing features to the doc, please do it preferably outside of feature PRs to allow the doc improvement to be backported.

**Comment by gares** [comment]
CI green modulo windows. If I get it right, this PR can be merged (and eventually another one improving the doc even more can be prepared)

**Comment by herbelin** [comment]
>  If I get it right, this PR can be merged

I believe so.

File: doc/changelog/03-notations/13265-master+allow-single-binder-entry.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  The :n:`binder` entry of :cmd:`Notation` can now be used in
```
Comment (Zimmi48) [inline doc/changelog/03-notations/13265-master+allow-single-binder-entry.rst:2]:
```suggestion
  The :n:`@binder` entry of :cmd:`Notation` can now be used in
```

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 4]
```diff
@@ -709,6 +709,31 @@ Note also that in the absence of a ``as ident``, ``as strict pattern`` or
...
+It is also possible to rely on Coq's own syntax of parsers using the
+`binder` modifier as follows:
+
+.. coqtop:: in
+
+   Notation "'myforall' p , [ P , Q ] " := (forall p, P -> Q)
+     (at level 200, p binder).
+
+In this case, all of `@ident`, `{@ident}`, `[@ident]`, `@ident:@type`,
+`{@ident:@type}`, `[@ident:@type]`, `'pattern` can be used in place of
```
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:724]:
```suggestion
In this case, all of :n:`@ident`, :n:`{@ident}`, :n:`[@ident]`, :n:`@ident:@type`,
:n:`{@ident:@type}`, :n:`[@ident:@type]`, :n:`'@pattern` can be used in place of
```

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 5]
```diff
@@ -709,6 +709,31 @@ Note also that in the absence of a ``as ident``, ``as strict pattern`` or
...
+the corresponding notation variable. In particular, the binding can be
+used to declare implicit arguments:
+
+.. coqtop:: all
+
+   Check fun (f : myforall {a}, [a=0, Prop]) => f eq_refl.
+   Check myforall '((x,y):nat*nat), [ x = y, True ].
+
+By using instead `closed binder`, parentheses, the same list of syntax
```
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:733]: I don't understand this ", parentheses,"

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 6]
```diff
@@ -709,6 +709,31 @@ Note also that in the absence of a ``as ident``, ``as strict pattern`` or
...
+for binders would be allowed except that `@ident:@type` would require
```
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:734]:
```suggestion
for binders would be allowed except that :n:`@ident:@type` would require
```

**Comment by ejgallego** [comment]
Needs a rebase, so it may be a good opportunity to address @Zimmi48 comments.

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 5 above)
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: A typo, sorry. I reworded the paragraph, as well as the first one.

**Comment by gares** [comment]
ci green

**Comment by ejgallego** [comment]
@coqbot: merge now
